### PR TITLE
53 profile details

### PR DESCRIPTION
**Pull request by tyxdamien**
created router for the profileDetails page and the nested router children to navigate between different tabs. Created profileTab Component and the three respective component cards - ProfileDetailCard, PaymentMethodsCard, ChangePasswordCard. Included the credit card type and the profile type. Also included mockedaData for the PaymentMethodsCard, can ignore MockProfileData.

### CODE REVIEW

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+                                    <Grid container direction="row" spacing={2}>
+                                        <Grid item>
+                                            <input
+                                                type="file"
+                                                accept="image/*"
+                                                id="profile-picture-upload"
+                                                style={{ display: "none" }}
+                                                onChange={handleFileUpload}
+                                            />
+                                            <label htmlFor="profile-picture-upload">
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:141]: Don't create an upload button, use the Avatar as an image input directly. Might need to use external libraries to make the Avatar can serve both as current pfp display and input as well

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+                                                <Button
+                                                    variant="contained"
+                                                    component="span"
+                                                >
+                                                    Upload
+                                                </Button>
+                                            </label>
+                                        </Grid>
+                                        <Grid item>
+                                            {profilePicture && (
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:151]: Display button still even if no `profilePicture`, only passed it to be disabled will do

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+                                        <Grid
+                                            container
+                                            direction="column"
+                                            gap={3}
+                                            justifyContent="flex-start"
+                                            alignItems="flex-start"
+                                            alignContent="flex-start"
+                                        >
+                                            <Grid item>
+                                                <TextField
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:183]: No need id, type is `text`. Read the API documentations for TextField

Comment (Bk49) [reply frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:183]: And btw, I have already created a `TextField` component for our app, please import the one under `components/common/form/TextField.tsx`, not the one by MUI

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+                                                                ) =>
+                                                                    setError(
+                                                                        newError
+                                                                    )
+                                                                }
+                                                                slotProps={{
+                                                                    textField: {
+                                                                        helperText:
+                                                                            errorMessage,
+                                                                    },
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:283]: Create a resuable DateField components in `components/common/form` directory so that it can be something reusable. Let me know if u face any struggles with this

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+                                            </Grid>
+                                        </Grid>
+                                    </Grid>
+                                    <Grid item xs={5.8}>
+                                        <Grid
+                                            container
+                                            direction="column"
+                                            gap={3}
+                                            justifyContent="flex-start"
+                                            alignItems="flex-start"
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:335]: What happens if I remove these? Does it really change everything?
Rmb to not declare properties that are by default already assigned by MUI

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:1]: This file is absolute madness with 400+ lines, I think there are a lot of fields that can be declared their respective `name`, `label`, `rules` etc and then `.map` them. Make use of `Grid`'s capability to render each field as different sizes based on the screen size and perhaps try putting other components such as the avatar in another file

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+                                justifyContent="flex-end"
+                                spacing={3}
+                                sx={{ mt: "1.5rem", mb: "1.5rem" }}
+                            >
+                                <Grid item>
+                                    <Button
+                                        type="button"
+                                        variant="contained"
+                                        onClick={handleCancel}
+                                        sx={{
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:458]: Check the `Button` api provided by MUI, there should be a variant that can help with the cancel type of button so you need not to declare the bg color and hover effects. (Currently, hovering it makes it turn purple...)

File: frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+                        <Grid item sx={{ mt: "1.5em" }}>
+                            <Divider />
+                        </Grid>
+                        <Grid item>
+                            <Grid
+                                container
+                                direction="row"
+                                justifyContent="flex-end"
+                                spacing={3}
+                                sx={{ mt: "1.5rem", mb: "1.5rem" }}
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/ProfileDetailsCard.tsx:451]: Make use of `my` for `mt`+`mb`

File: frontend/src/components/customer/ProfileDetailsCard.tsx/CountryData.ts [hunk 8]
```diff
@@ -0,0 +1,433 @@
+interface CountryType {
```
Comment (Bk49) [inline frontend/src/components/customer/ProfileDetailsCard.tsx/CountryData.ts:1]: You do not need this type, TypeScript can infer what is contained in `countriesData` without needing this type. Unless you need to export it, but I doubt so....

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:1]: 353 lines file....

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+    const [checked, setChecked] = useState<number[]>([]);
+
+    const [newPayment, setNewPayment] = useState<creditcard>({
+        card_number: "",
+        card_name: "",
+        cvv: "",
+        expiration_date: "",
+    });
+
+    const [formErrors, setFormErrors] = useState<Record<string, string>>({});
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:27]: Make use of `useForm` hook for errors

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx
Comment (Bk49) [reply frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:1]: Btw, this file needs to be fully revamped to make use of useForm hook, have a look at how I did it back in HEAP to not need all the unnecessary code for taking in form data

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        }
+    };
+    const handleCancel = () => {
+        setNewPayment({
+            card_number: "",
+            card_name: "",
+            cvv: "",
+            expiration_date: "",
+        });
+    };
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:120]: This is why we use the `useForm` hook, to avoid all the manual validation algorithm you write here. How confident are you that your algorithm always work? When the user go crazy on the inputs and test out all edge cases, you will likely have some parts that you didn't secure well...

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+    return (
+        <form onSubmit={handleSubmit}>
+            <Paper elevation={2} sx={{ mt: "3rem", mb: "3rem" }}>
+                <Grid container direction="column" gap={2}>
+                    <Grid container direction="column">
+                        <Grid item sx={{ ml: "2rem", mr: "2rem", mt: "2rem" }}>
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:126]: Same problem, make use and `mx` and `my`. You can even use `margin:"2rem 2rem 2rem 0"` if I am not wrong

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+                            <Heading
+                                variant="h2"
+                                color="secondary"
+                                style={{ marginLeft: "1rem" }}
+                            >
+                                Saved Payment Methods
+                            </Heading>
+                        </Grid>
+                        <Grid item sx={{ mt: "-2rem", ml: "2rem", mr: "2rem" }}>
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:135]: When you see yourself doing negative margins, it means some part in the code, you didn't optimize things very well. negative margins are generally a bad way to remove margins. Remove the margins on the parents that contributed to it instead of adding negative margins in the child components

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+                            <CheckboxList
+                                data={paymentsData}
+                                onToggle={handleToggle}
+                                checked={checked}
+                            />
+                        </Grid>
+                        <Grid item sx={{ ml: "2rem", mr: "2rem" }}>
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:142]: `mx` again

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+                                        color="error"
+                                        onClick={handleRemovePayment}
+                                    >
+                                        Remove Payment Methods
+                                    </Button>
+                                </Grid>
+                            </Grid>
+                        </Grid>
+                    </Grid>
+                    <Grid item sx={{ mt: "2rem", ml: "2rem", mr: "2rem" }}>
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:161]: Same margin problem again..

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+                                                    card_number: e.target.value,
+                                                })
+                                            }
+                                            sx={{
+                                                "& .MuiFormHelperText-root": {
+                                                    fontSize: "1rem",
+                                                    fontWeight: "400",
+                                                },
+                                            }}
+                                        />
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:202]: Import our TextField, not the one by MUI

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+                                                        e.target.value,
+                                                })
+                                            }
+                                        />
+                                    </Grid>
+                                </Grid>
+                            </Grid>
+                        </Grid>
+                    </Grid>
+                    <Grid item sx={{ ml: "2rem", mr: "2rem" }}></Grid>
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:315]: Is this extra?

File: frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+                                <Button
+                                    type="button"
+                                    variant="contained"
+                                    onClick={handleCancel}
+                                    sx={{
+                                        backgroundColor: "#000000",
+                                        opacity: "38%",
+                                    }}
+                                >
+                                    cancel
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/PaymentMethodsCard.tsx:337]: Should be a cancel associated variant of the button

File: frontend/src/components/customer/PaymentMethodsCard/CheckBoxList.tsx [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+                            onClick={() => onToggle(index)}
+                            dense
+                        >
+                            <ListItemIcon>
+                                <Checkbox
+                                    edge="start"
+                                    checked={checked.indexOf(index) !== -1}
+                                    tabIndex={-1}
+                                    disableRipple
+                                    inputProps={{ "aria-labelledby": labelId }}
```
Comment (Bk49) [inline frontend/src/components/customer/PaymentMethodsCard/CheckBoxList.tsx:38]: Create a custom `Checkbox` component. You can refer to how I did it in HEAP

File: frontend/src/components/customer/ChangePasswordCard/ChangePasswordCard.tsx
Comment (Bk49) [inline frontend/src/components/customer/ChangePasswordCard/ChangePasswordCard.tsx:1]: Same problem as the other comments given for everything here

**Review by Bk49 — CHANGES_REQUESTED** [review]
Please make the following changes and commit again, summary of changes
- Make use of the `mx` and `my` for your margin, the best is use `gap` whenever possible so you don't need to assign margins in every single components
- Read and use the `useForm` hook properly. You will need to use the `Controller` component provided by the API and `useForm` & `useFormContext`. 
- Whenever your file lines are longer than 100 lines, there is a problem already. Look for opportunities to map elements and create sub-components instead. This can avoid a lot of possible issues you will face when you keep writing boiler plate code (You might miss a property or two by accident)
- I haven't really check in detail how your app reacts when the screen shrinks, but I pointed some I noticed ady. Please try shrinking your screen and see for yourself if you as a mobile user is able to use the interface or not.

**Review by Bk49 — APPROVED** [review]
